File: leetcode/solution_3.py
'''
Summary

Attempt #1
Your own answer?: No
Reference: https://leetcode.com/problems/longest-substring-without-repeating-characters/discuss/1731/A-Python-solution-85ms-O(n)
Runtime: 48 ms, faster than 96.47% of Python3 online submissions for Longest Substring Without Repeating Characters.
Memory Usage: 14.4 MB, less than 55.31% of Python3 online submissions for Longest Substring Without Repeating Characters.

Memo: I also came up with the idea of using one pointer pointing to the head of the range and another pointing to the iterater variable while keeping track of each value's latest index, 
but somehow failed to get the size out of the custom range.
'''

import sys
input = sys.stdin.readline

class Solution:
    def __init__(self) -> None:
        s = "abcabcbb"
        print(self.lengthOfLongestSubstring(s))

    def lengthOfLongestSubstring(self, s: str) -> int:
        start = maxLength = 0
        used = {}

        for idx, val in enumerate(s):
            if val in used and start <= used[val]:
                start = used[val] + 1
            else:
                maxLength = max(maxLength, idx - start + 1)
        
            used[val] = idx

        return maxLength

Solution()

File: leetcode/solution_3.cs
// Time Complexity: O(n) where n is the length of the given string
// Space Complexity: O(n) where n is the length of the given string


public class Solution {
    public int LengthOfLongestSubstring(string s) {
        var start = 0;
        var end = 0;
        var localLength = 0;
        var globalLength = 0;
        var hashMap = new Dictionary<char, int>();
        
        while (end < s.Length)
        {
            var c = s[end];
            if (hashMap.TryGetValue(c, out int index) && index != -1)
            {
                globalLength = Math.Max(globalLength, localLength);
                localLength = 0;
                while (start < end)
                {
                    hashMap[s[start]] = -1;
                    start++;
                }
                start = index + 1;
                end = start;
                continue;
            }
            
            hashMap[c] = end;
            localLength++;
            end++;
        }
        
        return Math.Max(globalLength, localLength);
    }
}